### PR TITLE
AITL migration

### PR DESCRIPTION
**Pull request by gjolly**
Penguinator has been deprecated and Azure introduced a new API called Azure Image Testing
for Linux (AITL). It also uses LISA in the backend.

### CODE REVIEW

File: README.md
Comment (j5awry) [inline README.md]: jobtemplate as one word feels odd. I won't block on naming, but `job-template` or just `template` may be better

**Review by j5awry — APPROVED** [review]
approving with one optional nit

File: .gitignore
Comment (danpdraper) [inline .gitignore]: Ignore statements that reflect developer workstation preferences should reside in developer-workstation-level `.gitignore` files.

File: README.md
Comment (danpdraper) [inline README.md]: The first instance of `API` is redundant.

File: README.md
Comment (danpdraper) [inline README.md]: The references to `penguinator` littered throughout this repository will undoubtedly prove confusing in the future. Can we change the names of the repository and snap?

File: README.md
Comment (danpdraper) [reply README.md]: Agreed. Either `*-job-template(s)` or `*-template(s)`.

File: cli/main.py
Comment (danpdraper) [inline cli/main.py]: `kwargs` should always contain a `"resource_group"` key. If `kwargs` does not contain a `"resource_group"` key, the application should at least signal that something is awry. As such, consider replacing `kwargs.get("resource_group")` with `kwargs["resource_group"]`.

File: cli/main.py
Comment (danpdraper) [inline cli/main.py]: The help strings produced by this script start with a mix of lowercase and uppercase letters. Please choose one approach and apply across the board.

File: cli/main.py
Comment (danpdraper) [inline cli/main.py]: You could reduce this to: `template["region"] = regions if regions else []`. A similar comment applies to lines 83-86.

File: cli/main.py
Comment (danpdraper) [inline cli/main.py]: It would likely be helpful to have some kind of preamble in these two print statements to assist those trying to interpret the script's output.

File: cli/main.py
Comment (danpdraper) [inline cli/main.py]: Looks like this should be a choice with a default.

File: cli/main.py
Comment (danpdraper) [inline cli/main.py]: I take it that the order of the decorators need not match the order of the function parameters?

File: cli/main.py
Comment (danpdraper) [inline cli/main.py]: It would be helpful for the error to include the value that the function believes to be the URN.

File: snap/snapcraft.yaml
Comment (danpdraper) [inline snap/snapcraft.yaml]: This URL seems to be invalid.

**Review by danpdraper — CHANGES_REQUESTED** [review]

File: README.md
Comment (gjolly) [reply README.md]: Can we do this later please? It will require a lot of changes: repo, python package, CLI, snap and internal tools.

File: cli/main.py
Comment (gjolly) [reply cli/main.py]: I disagree with this, having the output be the json object returned by the API makes it very easy to use this CLI command in automated jobs (which is its main purpose).
Ideally the user should have the ability to format the output the way they'd like but it's out of scope for this PR.

File: cli/main.py
Comment (gjolly) [reply cli/main.py]: Indeed, it's based on names: https://click.palletsprojects.com/en/8.1.x/options/#name-your-options

File: README.md
Comment (gjolly) [reply README.md]: Actually, I did it. I will rename the repo once this is merged.

File: cli/main.py
Comment (danpdraper) [reply cli/main.py]: I'm not saying that you should do away with the JSON,  but rather that the JSON should be preceded by some simple text that explains what it is, e.g.:

```
print(f"Response from AITL:\n{json.dumps(resp.json(), indent=2)}")
```

File: cli/main.py
Comment (danpdraper) [reply cli/main.py]: This still needs to be addressed.

File: cli/main.py
Comment (gjolly) [reply cli/main.py]: My point is that it will break `jq`. Or if you want to do that we should do it on stderr instead of stdout.

File: README.md
Comment (AleksaSvitlica) [inline README.md:47]: Will the repository be renamed as part of this change?

File: README.md
Comment (AleksaSvitlica) [inline README.md:3]: Will you create a new snapcraft package for this?

File: cli/main.py
Comment (AleksaSvitlica) [inline cli/main.py]: Could you clarify this option? The flag makes it sound like a singular name but the help text implies multiple.

File: cli/main.py
Comment (AleksaSvitlica) [inline cli/main.py]: Does this need to be reversed? What does that accomplish?

File: cli/main.py
Comment (AleksaSvitlica) [inline cli/main.py]: Could you add type hints here?

File: cli/main.py
Comment (AleksaSvitlica) [inline cli/main.py]: I do wonder if these common options would be better handled by creating a Click group. That does introduce a change to the CLI though since would have to call it like 'aitl-cli <group> <command>'.

**Review by AleksaSvitlica — COMMENTED** [review]
Nothing blocking but some comments and questions.

File: cli/main.py
Comment (danpdraper) [reply cli/main.py]: Ah, okay. Leave it as is.

Comment (danpdraper) [reply cli/main.py]: I don't have the option to resolve conversations in this pull request. Do I have [write access to this repository](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#resolving-conversations)?

File: cli/main.py
Comment (danpdraper) [reply cli/main.py]: Sounds good. Please resolve this conversation.

**Review by danpdraper — APPROVED** [review]

File: cli/main.py
Comment (gjolly) [reply cli/main.py]: Indeed. You should have access now.

File: README.md
Comment (gjolly) [reply README.md:3]: yes

File: README.md
Comment (gjolly) [reply README.md:47]: yes it is now.

File: cli/main.py
Comment (gjolly) [reply cli/main.py]: Yeah I think it would be better, I was just reluctant to do a full rewrite of those commits. I think I will end up doing this.

File: .gitignore
Comment (danpdraper) [inline .gitignore]: `credentials` (typo)

File: README.md
Comment (AleksaSvitlica) [reply README.md:47]: What do you mean by it is now? I get a 404 at the new url.

File: pyproject.toml
Comment (danpdraper) [inline pyproject.toml]: This formulation suggests to me that something named "Basic Azure Image Testing" is for "Linux API client". Consider removing `"Basic"` and adding the acronym in parentheses after `"Linux"`, i.e., `"Azure Image Testing for Linux (AITL) API client"`.

File: src/cli/main.py
Comment (AleksaSvitlica) [inline src/cli/main.py]: I am of the opinion this should be --priorities and --case-names given that they are lists. 

Alternatively consider changing these to the multiple=True options. Looking at how this gets used in `_create_template` they get split into lists anyways. If you expect a lot of case names to be passed then maybe this is undesirable as you'll have a lot of extra characters needed when calling the cli.

**Review by AleksaSvitlica — APPROVED** [review]
One question and one suggestion but nothing blocking. Looks good to me.

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `--priorities` (plural)
`test priorities` (plural)

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `--case-names` (plural)
`test cases` (plural)

**Comment by gjolly** [comment]
@AleksaSvitlica the repo is not there yet because I want to rename this one as soon as this PR gets merged.

File: src/cli/main.py
Comment (gjolly) [reply src/cli/main.py]: Good point

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py:140]: In the interest of consistency with `--priorities` and `--case-names`, should this parameter also take a comma-separated list?

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `--priorities and --case-names`

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: As above, should this parameter also take a comma-separated list in the interest of consistency with `--priorities` and `--case-names`?

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `--priorities` (plural)
`test priorities` (plural)

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `--case-names` (plural)
`test cases` (plural)

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py:223]: This parameter is defined twice (see lines 209-211).

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `--priorities and --case-names`

**Review by danpdraper — CHANGES_REQUESTED** [review]

File: src/cli/main.py
Comment (gjolly) [reply src/cli/main.py]: I will do the opposite as suggested by Aleksa and make the two others use `multiple`.

**Comment by gjolly** [comment]
So, I changed the name of those options for something more comprehensible but I did not use the plural because I now use multiple=True. HOWEVER, I pass a third parameter to click.option so that the *variable names* are now plural :tada:

**Comment by gjolly** [comment]
BTW I also tested the CLI manually to make sure I can create and get jobs (since there is no test).

File: pyproject.toml
Comment (danpdraper) [inline pyproject.toml]: The `AITL` acronym does not include `API` and thus should precede `API` (i.e., `...(AITL) API`).

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `Test priorities to run` (plural)

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `Test cases to run` (plural)

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `Test priorities to run` (plural)

File: src/cli/main.py
Comment (danpdraper) [inline src/cli/main.py]: `Test cases to run` (plural)

**Review by danpdraper — CHANGES_REQUESTED** [review]

File: src/cli/main.py
Comment (danpdraper) [reply src/cli/main.py:223]: This still needs to be addressed.

File: src/cli/main.py
Comment (gjolly) [reply src/cli/main.py]: Really it depends how you read it. It will look like this
```
aitl-cli --case-name xxx --case-name yyy --case-name zzz
```
So it should be singular IMO

Comment (danpdraper) [reply src/cli/main.py]: What does the help output for the `--case-name` parameter look like though?

Comment (danpdraper) [reply src/cli/main.py]: Is it something like `--case-name <first_case> [--case-name <second_case>] [--case-name <third_case>]...`? If yes, then I believe `"Test cases"` would be appropriate.

Comment (gjolly) [reply src/cli/main.py]: It looks like this:

```bash
❯ aitl-cli create-job --help
Usage: aitl-cli create-job [OPTIONS]

  Create a new test job

  If --test-priority and --test-case are not set, only p0 tests will be run
  (smoke tests).

Options:
  -n, --name TEXT                 Job name.  [required]
  -u, --marketplace-image-urn TEXT
                                  URN of the image ("az vm image list -p
                                  Canonical --all").
  -v, --vhd-sas-url TEXT          SAS URL of a VHD to test.
  -a, --architecture [x64|arm64]  Architecture of the image.
  -g, --vm-generation [1|2]       Hyper-V generation.
  -t, --template-name TEXT        Job template name.
  -n, --name TEXT                 Job template name.  [required]
  -s, --vm-size TEXT              VM size.
  -p, --test-priority TEXT        Test priority to run.
  -c, --test-case TEXT            Test case to run.
  -l, --location TEXT             Job template location, it's not required to
                                  be changed.
  -r, --region TEXT               Regions where the test resources will be
                                  provisioned.
  --help                          Show this message and exit.
```

Comment (danpdraper) [reply src/cli/main.py]: Thanks for clarifying. I agree that singular is appropriate.

File: pyproject.toml
Comment (danpdraper) [reply pyproject.toml]: This still needs to be addressed.

**Review by danpdraper — APPROVED** [review]

**Pull request merged**
